rover_shared_libs/rover_can2/src/rover_can2/drivers/driver_linux.hpp

```diff
@@ -1,17 +1,322 @@
 #ifndef DRIVER_LINUX_HPP
 #define DRIVER_LINUX_HPP
 
+#include "rover_can2/can_msg.hpp"
+#include "rover_can2/constant.hpp"
 #include "rover_can2/drivers/driver_base.hpp"
 
+#include "rover_lib2/helpers/log.hpp"
+#include "rover_lib2/helpers/circular_buffer.hpp"
+#include "rover_lib2/helpers/watchdog.hpp"
+#include "rover_lib2/helpers/time.hpp"
+
+#include <string>
+#include <thread>
+#include <chrono>
+#include <fcntl.h>
+#include <linux/can.h>
+#include <linux/can/raw.h>
+#include <sys/socket.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+DEFINE_LOG_NODE(DriverLinux, Logger::eNodeState::ON);
+
 namespace RoverCan2::Drivers
 {
     class DriverLinux : public DriverBase<DriverLinux>
     {
+        static constexpr auto INTERFACE_STABILIZATION_TIME_MS = 1500;
+
+        enum class eState : size_t
+        {
+            UNINSTALLED,
+            RUNNING,
+            TX_QUEUE_FULL
+        };
+
       public:
-        DriverLinux() = default;
+        DriverLinux(const std::string& interfaceName_ = "canRovus"):
+            _interfaceName(interfaceName_),
+            _state(eState::UNINSTALLED),
+            _recvWatchdog(2ULL * 1'000ULL / static_cast<uint64_t>(Constant::MASTER_HEARTBEAT_RATE_HZ))
+        {
+        }
+
+        ~DriverLinux()
+        {
+            cleanupCanSocket();
+        }
+
+        // init() is useless in this case
+        void __init()
+        {
+            LOG_INFO(Logger::Nodes::DriverLinux, "Initializing CAN linux driver");
+            this->createCanSocket();
+        }
+
+        void __update(void)
+        {
+            if (!_recvWatchdog.isOk())
+            {
+                LOG_DEBUG(Logger::Nodes::DriverLinux,
+                          "Receive watchdog timeout: no CAN message received within the expected interval");
+            }
+
+            switch (_state)
+            {
+                case eState::UNINSTALLED:
+                    createCanSocket();
+                    break;
+                case eState::RUNNING:
+                    [[fallthrough]];
+                case eState::TX_QUEUE_FULL:
+                    this->processNewMessage();
+                    break;
+            }
+        }
+
+        std::optional<CanMsg> _getMsg(void)
+        {
+            return _msgBuffer.getValue();
+        }
+
+        bool _sendMsg(const CanMsg& canMsg_)
+        {
+            if (_state < eState::RUNNING)
+            {
+                LOG_DEBUG(Logger::Nodes::DriverLinux, "Can't send msg, driver not initialized.");
+                return false;
+            }
+
+            if (_socket_fd < 0)
+            {
+                LOG_DEBUG(Logger::Nodes::DriverLinux, "Invalid socket file descriptor: %d", _socket_fd);
+                return false;
+            }
+
+            struct can_frame frame
+            {
+            };
+            frame.can_id = static_cast<uint32_t>(canMsg_.getCanID());
```

Comment:
This is true but might want to use the direct type of can_id:

```suggestion
            frame.can_id = static_cast<canid_t>(canMsg_.getCanID());
```